pkg/api/graph/graph.go

```diff
@@ -0,0 +1,397 @@
+package graph
+
+import (
+	"fmt"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/concrete"
+)
+
+type Node struct {
+	concrete.Node
+	UniqueName
+}
+
+type UniqueName string
+
+func (n UniqueName) UniqueName() string {
+	return string(n)
+}
+
+type uniqueNamer interface {
+	UniqueName() string
+}
+
+// UniqueNodeInitializer is a graph that allows nodes with a unique name to be added without duplication.
+// If the node is newly added, true will be returned.
+type UniqueNodeInitializer interface {
+	FindOrCreate(name UniqueName, fn NodeInitializerFunc) (graph.Node, bool)
+}
+
+type NodeInitializerFunc func(Node) graph.Node
+
+func EnsureUnique(g UniqueNodeInitializer, name UniqueName, fn NodeInitializerFunc) graph.Node {
+	node, _ := g.FindOrCreate(name, fn)
+	return node
+}
+
+type MutableDirectedEdge interface {
+	AddEdge(head, tail graph.Node, edgeKind int)
+}
+
+type MutableUniqueGraph interface {
+	graph.Mutable
+	MutableDirectedEdge
+	UniqueNodeInitializer
+}
+
+type Edge struct {
+	concrete.Edge
+	K int
```

Comment:
So it doesn't overlap with "Kind()" the method.  Consistent with underlying gonum/graph

----- Original Message -----
> > +	return node
> > +}
> > +
> > +type MutableDirectedEdge interface {
> > +	AddEdge(head, tail graph.Node, edgeKind int)
> > +}
> > +
> > +type MutableUniqueGraph interface {
> > +	graph.Mutable
> > +	MutableDirectedEdge
> > +	UniqueNodeInitializer
> > +}
> > +
> > +type Edge struct {
> > +	concrete.Edge
> > +	K int
> 
> K?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1314/files#r27434692